File: 06/solution.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void reset(int arr[]) {
    for (int i = 0; i < 26; ++i) {
        arr[i] = 0;
    }
}

int count_set(int arr[]) {
    int set = 0;
    for (int i = 0; i < 26; ++i) {
        if (arr[i] == 1) {
            ++set;
        }
    }

    // printf("Counted: %d\n", set);
    return set;
}

void print_arr(int arr[]) {
    for (int i = 0; i < 26; ++i)
    {
        printf("%d", arr[i]);

    }
    printf("\n");
}

int main(int argc, char **argv)
{
    FILE *fptr;
    char file[] = "input";
    char line[28];
    int answers_all[26] = {0};
    int answers_all_count = 0;
    int answers_any[26] = {0};
    int answers_any_count = 0;
    int these_answers[26];
    bool new_group = true;

    if ((fptr = fopen("input", "r")) == NULL)
    {
        printf("Unable to open file: %s\n", file);
        return 1;
    }

    while (fgets(line, sizeof line, fptr))
    {
        // printf("Line: %lu - %s\n", strlen(line), line);
        if (line[strlen(line) - 1] == '\n') {
            line[strlen(line) - 1] = '\0';
        }

        if (strlen(line) == 0) {
            answers_any_count += count_set(answers_any);
            answers_all_count += count_set(answers_all);
            // printf("Any: %d\nAll: %d\n", answers_any_count, answers_all_count);
            reset(answers_any);
            reset(answers_all);
            new_group = true;
        }
        else {
            int len = strlen(line);
            int idx;
            reset(these_answers);
            for (int i = 0; i < len; ++i)
            {
                idx = line[i] - 0x61;
                these_answers[idx] = 1;
            }

            // print_arr(answers_any);
            // print_arr(answers_all);
            for (int i = 0; i < 26; ++i)
            {
                answers_any[i] |= these_answers[i];

                if (new_group)
                {
                    answers_all[i] = these_answers[i];
                } else {
                    answers_all[i] &= these_answers[i];
                }

            }
            // print_arr(these_answers);
            // print_arr(answers_any);
            // print_arr(answers_all);

            if (new_group)
            {
                new_group = false;
            }
        }
    }

    answers_any_count += count_set(answers_any);
    answers_all_count += count_set(answers_all);

    printf("Any: %d\nAll: %d\n", answers_any_count, answers_all_count);

    return 0;
}

File: 06/solution.rs
use std::env;
use std::fs;
use std::collections::HashSet;
use std::io::{
    BufReader,
    BufRead,
};

fn main() -> Result<(), std::io::Error> {
    let args: Vec<String> = env::args().collect();
    let filename = &args[1];

    let file = fs::File::open(filename)?;
    let mut buf = BufReader::new(file);
    let mut line = String::new();
    let mut answers_any: HashSet<char> = HashSet::new();
    let mut answers_all: HashSet<char> = HashSet::new();
    let mut group_answer_any = 0;
    let mut group_answer_all = 0;
    let mut new_group = true;

    loop {
        match buf.read_line(&mut line) {
            Ok(bytes_read) => {

                if bytes_read == 0 {
                    // EOF
                    group_answer_any += answers_any.len();
                    group_answer_all += answers_all.len();
                    // println!("Counted: {}\nCounted: {}", answers_any.len(), answers_all.len());
                    break;
                }

                // remove newline
                line = line.trim().into();
                if !line.is_empty() {
                    let these_answers = line.chars().collect();

                    answers_any.extend(&these_answers);

                    if new_group {
                        answers_all.extend(&these_answers);
                        new_group = false;
                    } else {
                        answers_all = answers_all
                            .intersection(&these_answers)
                            .map(|c| *c)
                            .collect();
                    }
                } else {
                    group_answer_any += answers_any.len();
                    group_answer_all += answers_all.len();
                    // println!("Counted: {}\nCounted: {}", answers_any.len(), answers_all.len());
                    // println!("Any: {}\nAll: {}", group_answer_any, group_answer_all);
                    answers_any.clear();
                    answers_all.clear();
                    new_group = true;
                }

                line.clear();
            }
            Err(err) => {
                return Err(err);
            }
        }
    }

    println!("Any: {}\nAll: {}", group_answer_any, group_answer_all);

    Ok(())
}